docs/proposals/container-runtime-interface-v1.md

```diff
@@ -0,0 +1,182 @@
+high-level, declarative interface has caused higher integration and maintenance
+cost, and also slowed down feature velocity for the following reasons.
+  1. **Not every container runtime supports the concept of pods natively**.
+     When integrating with Kubernetes, a significant amount of work needs to
+     go into implementing a shim of significant size to support all pod
+     features. This also adds maintenance overhead (e.g., `docker`).
+  2. **High-level interface discourage code sharing and reuse among runtimes**.
+     E.g, each runtime implements an all-encompassing `SyncPod()` function that
+     (re-)starts pods/containers and manage lifecycle hooks.
+  3. **Pod Spec is susceptible to inconsistent interpretations**. E.g., `rkt`
+     does not support container-level operations and assume immutable pods.
+  4. **Pod Spec is evolving rapidly**. New features are being added constantly.
+     Any pod-level change or addition requires changing of all container
+     runtime shims. E.g., init containers and volume containers.
+
+## Goals and Non-Goals
+
+The goals of defining the interface are to
+ - **improve extensibility**: Easier container runtime integration.
+ - **improve feature velocity**
+ - **improve code maintainability**
+
+The non-goals include
+ - proposing *how* to integrate with new runtimes, i.e., where the shim
+   resides. The discussion of adopting a client-server architecture is tracked
+   by [#13768](https://issues.k8s.io/13768), where benefits and shortcomings of
+   such an architecture is discussed.
+ - adding support to Windows containers. Windows container support is a
+   parallel effort and is tracked by [#22623](https://issues.k8s.io/22623). The
+   new interface will leave room for future integration, but will
+ - re-defining Kubelet's internal interfaces. These interfaces, though, may affect
+   Kubelet's maintainability, is not relevant to runtime integration's.
+
+## Requirements
+
+ * Support existing integrated container runtimes
+ * Support hypervisor-based container runtimes
+
+## Container Runtime Interface
+
+The main idea of the proposal is to adopt an imperative container-level
+interface, which allows Kubelet to control the lifecycles of the containers
+inside a pod, across all container runtimes. A separate PodSandbox is concept
+is defined to represent the environment in which a group of containers in a
+pod will be created and run. The container runtimes may interpret the sandbox
+concept differently based on how it operates internally. For runtimes relying
+on hypervisor, sandbox represents a virtual machine. For others, it can be a
+container holding the namespaces.
+
+In short, a PodSandbox should have the following features.
+
+ * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
+   support additional security features.
+ * **Resource requirements**: A sandbox should implement pod-level resource
```

Comment:
`isolation` is the primary goal. We have various kinds of isolation. May be
think of a better name for namespaces/VMs?

On Mon, May 16, 2016 at 4:44 PM, Yu-Ju Hong <notifications@github.com>
wrote:

> In docs/proposals/container-runtime-interface-v1.md
> <https://github.com/yujuhong/kubernetes/pull/1#discussion_r63444960>:
>
> > +## Container Runtime Interface
> > +
> > +The main idea of the proposal is to adopt an imperative container-level
> > +interface, which allows Kubelet to control the lifecycles of the containers
> > +inside a pod, across all container runtimes. A separate PodSandbox is concept
> > +is defined to represent the environment in which a group of containers in a
> > +pod will be created and run. The container runtimes may interpret the sandbox
> > +concept differently based on how it operates internally. For runtimes relying
> > +on hypervisor, sandbox represents a virtual machine. For others, it can be a
> > +container holding the namespaces.
> > +
> > +In short, a PodSandbox should have the following features.
> > +
> > + * **Isolation**: E.g., Linux namespaces or a full virtual machine, or even
> > +   support additional security features.
> > + * **Resource requirements**: A sandbox should implement pod-level resource
>
> How about "resource specification"? Then we don't have to repeat
> "isolation".
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly or view it on GitHub
> <https://github.com/yujuhong/kubernetes/pull/1/files/ddd041c68ca94d95fa591dce703932fe5d577dd0#r63444960>
>